src/evm.rs

```diff
@@ -132,6 +168,88 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EvmNamespaceT
             }
         })
     }
+
+    fn snapshot(&self) -> BoxFuture<Result<U64>> {
+        let inner = Arc::clone(&self.node);
+        let snapshots = Arc::clone(&self.snapshots);
+
+        Box::pin(async move {
+            // validate max snapshots
+            snapshots
+                .read()
+                .map_err(|err| {
+                    log::error!("failed acquiring read lock for snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })
+                .and_then(|snapshots| {
+                    if snapshots.len() >= MAX_SNAPSHOTS as usize {
+                        log::error!("maximum number of '{}' snapshots exceeded", MAX_SNAPSHOTS);
+                        Err(into_jsrpc_error(Web3Error::InternalError))
+                    } else {
+                        Ok(())
+                    }
+                })?;
+
+            // snapshot the node
+            let snapshot = inner
+                .read()
+                .map_err(|err| {
+                    format!("failed acquiring read lock to node for snapshot: {:?}", err)
+                })
+                .and_then(|reader| reader.snapshot())
+                .map_err(|err| {
+                    log::error!("failed creating snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })?;
+            snapshots
+                .write()
+                .map(|mut snapshots| {
+                    snapshots.push(snapshot);
+                    snapshots.len()
+                })
+                .map_err(|err| {
+                    log::error!("failed storing snapshot: {:?}", err);
+                    into_jsrpc_error(Web3Error::InternalError)
+                })
+                .map(U64::from)
+        })
+    }
+
+    fn revert_snapshot(&self, snapshot_id: U64) -> BoxFuture<Result<bool>> {
+        let inner = Arc::clone(&self.node);
+        let snapshots = Arc::clone(&self.snapshots);
+
+        Box::pin(async move {
+            let mut snapshots = snapshots.write().map_err(|err| {
+                log::error!("failed acquiring read lock for snapshots: {:?}", err);
+                into_jsrpc_error(Web3Error::InternalError)
+            })?;
+            let snapshot_id_index = snapshot_id.as_usize().saturating_sub(1);
+            if snapshot_id_index >= snapshots.len() {
```

Comment:
nit/suggestion: We can also check for negative values here to avoid potential errors down below

Reply:
`saturating_sub` should clip the operation at `0`. It's pretty much impossible to get to the negative range or underflow, since we're dealing with unsigned arithmetic and `U64` request deserializers.